Add codespell as a linter

benfred opened the pull request:
Similar to https://github.com/rapidsai/cudf/pull/12097, this adds codespell as a linter to the pre-commit config, and fixes various spelling errors it highlights. (https://github.com/codespell-project/codespell)

csadorf commented on python/cuml/tests/test_linear_svm.py:
```diff
@@ -41,15 +41,15 @@
 def good_enough(myscore: float, refscore: float, training_size: int):
     myerr = 1.0 - myscore
-    referr = 1.0 - refscore
+    refer = 1.0 - refscore
```
Please revert the changes to this module. "Referr" is short for "reference error".

benfred replied:
done

csadorf commented on cpp/src/tsne/cannylab/bh.cu:
```diff
@@ -984,4 +984,4 @@ int main(int argc, char* argv[])
   return 0;
-}
\ No newline at end of file
+}
```
Unrelated change?

csadorf commented on pyproject.toml:
```diff
@@ -0,0 +1,9 @@
+# ignore short words, and typename parameters like OffsetT
+ignore-regex = "\\b(.{1,4}|[A-Z]\\w*T)\\b"
+ignore-words-list = "inout,numer,startd,couldn"
+builtin = "clear"
+quiet-level = 3
```
It's not immediately obvious to me what these config options do, maybe you can add comments there as well?

csadorf reviewed (changes requested):
Very nice! I like that we can use this pre-commit hook to more easily avoid these kind of common mistakes in the future.

I went through all changes and found a few spurious and wrong ones, I would appreciate if you could address those.

Further, I think it would be good to add a summary of the hooks' purpose to the contributor guide [here](https://github.com/rapidsai/cuml/blob/branch-23.04/CONTRIBUTING.md#summary-of-pre-commit-hooks). It would also be great to have a brief note with guidance on how to disable the check for specific words or lines.

csadorf reviewed (approved):
Great, thanks for addressing the review comments so quickly. It would still be nice if you could add a note about how to ignore certain suggestion (`# noqa` ?) to the contributor guide, otherwise LGTM!

benfred commented:
> Great, thanks for addressing the review comments so quickly. It would still be nice if you could add a note about how to ignore certain suggestion (`# noqa` ?) to the contributor guide, otherwise LGTM!

thanks @csadorf ! I've added some text about how to ignore suggestions in the last commit

csadorf commented:
/merge

vyasr commented:
/merge
